presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -506,6 +507,121 @@ private static Domain extractEquatableDomain(ComparisonExpression.Type compariso
+                        return new ComparisonExpression(GREATER_THAN, fieldReference, valueExpression);
+                    }
+                }
+                case LESS_THAN_OR_EQUAL:
+                case LESS_THAN: {
+                    if (coercedValueIsLessThanOriginal) {
+                        return new ComparisonExpression(LESS_THAN_OR_EQUAL, fieldReference, valueExpression);
+                    }
+                    else if (coercedValueIsEqualToOriginal) {
+                        return new ComparisonExpression(comparisonType, fieldReference, valueExpression);
+                    }
+                    else if (coercedValueIsGreaterThanOriginal) {
+                        return new ComparisonExpression(LESS_THAN, fieldReference, valueExpression);
+                    }
+                }
+                case EQUAL: {
+                    if (coercedValueIsEqualToOriginal) {
+                        return new ComparisonExpression(EQUAL, fieldReference, valueExpression);
+                    }
+                    // Return something that is false for all non-null values
+                    return and(new ComparisonExpression(EQUAL, fieldReference, valueExpression),
+                            new ComparisonExpression(NOT_EQUAL, fieldReference, valueExpression));
+                }
+                case NOT_EQUAL: {
+                    if (coercedValueIsEqualToOriginal) {
+                        return new ComparisonExpression(comparisonType, fieldReference, valueExpression);
+                    }
+                    // Return something that is true for all non-null values
+                    return or(new ComparisonExpression(EQUAL, fieldReference, valueExpression),
+                            new ComparisonExpression(NOT_EQUAL, fieldReference, valueExpression));
+                }
+                case IS_DISTINCT_FROM: {
+                    if (coercedValueIsEqualToOriginal) {
+                        return new ComparisonExpression(comparisonType, fieldReference, valueExpression);
+                    }
+                    return TRUE_LITERAL;
+                }
+            }
+
+            throw new IllegalArgumentException("Unhandled type: " + comparisonType);
+        }
+
+        private Optional<Object> floorValue(Type fromType, Type toType, Object value)
+        {
+            Optional<Signature> saturatedFloorCastOperator = getSaturatedFloorCastOperator(fromType, toType);
+            if (!saturatedFloorCastOperator.isPresent()) {
+                return Optional.empty();
+            }
+            return Optional.of(functionInvoker.invoke(saturatedFloorCastOperator.get(), session.toConnectorSession(), value));
+        }
+
+        private Optional<Signature> getSaturatedFloorCastOperator(Type fromType, Type toType)
+        {
+            if (!metadata.getFunctionRegistry().canResolveOperator(SATURATED_FLOOR_CAST, toType, ImmutableList.of(fromType))) {
+                return Optional.empty();
```

Comment:
I have used `map` in every other places that you noticed, but it can't be used here, cause there are no method in `FunctionRegistry` that return `Optional`.